src/python/pants/engine/goal.py

```diff
@@ -109,16 +109,62 @@ class _GoalOptions(object):
   """A marker trait for the anonymous inner `Goal.Options` classes for `Goal`s."""
 
 
+class Outputting:
+  """A mixin for Goal that adds Options to support the `output` context manager.
+
+  Allows output to go to a file or to stdout.
+
+  Useful for goals that write non-line-oriented output, such as JSON.
+  """
+
+  @classmethod
+  def register_options(cls, register):
+    super().register_options(register)
+    register('--output-file', metavar='<path>',
+             help='Write output to this file.  If unspecified, writes to stdout.')
+
+  @classmethod
+  @contextmanager
+  def output(cls, output_options, console):
+    """Given Goal.Options and a Console, yields a function for writing output.
+
+    The passed options instance will generally be the `Goal.Options` of a `Outputting` `Goal`.
+    """
+    if type(output_options) != cls.Options:
+      raise AssertionError(
+          'Expected Options for `{}`, got: {}'.format(cls.__name__, output_options))
+
+    stdout_file = None
+
+    if output_options.values.output_file:
+      stdout_file = open(output_options.values.output_file, 'w')
+      write_stdout = lambda msg: stdout_file.write(msg)
+    else:
+      write_stdout = lambda msg: console.write_stdout(msg)
+
+    try:
+      yield write_stdout
+    finally:
+      if stdout_file:
+        stdout_file.close()
+      console.flush()
+
+
 class LineOriented:
```

Comment:
Should this extend `Outputting` to get the `--output-file` option, and re-use the logic? It looks like there is a lot copied here.

Reply:
See my comments on the PR. I started down that route, but it got a bit complex due to the distinction between print() and write(). But I also wonder if we need `LineOriented` at all? The only thing it adds to the mix is `--sep`, which seems unnecessary? I can't think of another tool that lets you pick the line end separator...

Reply:
See John's `\0` example.

But more generally, if we'd like to deprecate `--sep`, we should do it in v1 as well: this mixin was intended to allow goals to emulate v1 options.